File: 2121-FindIfPathExistsInGraph/2121-FindIfPathExistsInGraph.java
// Last updated: 8/12/2026, 11:47:13 AM
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        
        // Create adjacency list
        ArrayList<Integer>[] graph = new ArrayList[n];
        
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }

        // Add edges (bidirectional)
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];

            graph[u].add(v);
            graph[v].add(u);
        }

        boolean[] visited = new boolean[n];

        return dfs(source, destination, graph, visited);
    }

    private boolean dfs(int node, int destination, ArrayList<Integer>[] graph, boolean[] visited) {
        
        if (node == destination) {
            return true;
        }

        visited[node] = true;

        for (int neighbour : graph[node]) {
            if (!visited[neighbour]) {
                if (dfs(neighbour, destination, graph, visited)) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: 2121-FindIfPathExistsInGraph/2121-FindIfPathExistsInGraph.py
# Last updated: 8/12/2026, 11:47:06 AM
import collections
from typing import List

class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        if source == destination:
            return True

        # Build the adjacency list
        graph = collections.defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)

        # Use a stack for DFS and a set for visited nodes
        stack = [source]
        visited = {source}

        while stack:
            current_node = stack.pop()

            if current_node == destination:
                return True

            for neighbor in graph[current_node]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    stack.append(neighbor)
        
        return False